### PR TITLE
podman port

### PR DESCRIPTION
**Pull request by baude**
podman port reports the port mappings per container.  it can be used
to report the ports ofa single container or latest container or all
containers.

in the case of a single container, the user can add an option filter for
port and protocol.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

File: cmd/podman/port.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+	"github.com/projectatomic/libpod/libpod"
+	"strings"
```
Comment (rhatdan) [inline cmd/podman/port.go]: BAD IDE

File: cmd/podman/port.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+		Usage:       "List port mappings or a specific mapping for the container",
+		Description: portDescription,
+		Flags:       portFlags,
+		Action:      portCmd,
+		ArgsUsage:   "CONTAINER-NAME [mapping]",
+	}
+)
+
+func portCmd(c *cli.Context) error {
+	var userProto, containerName string
```
Comment (rhatdan) [inline cmd/podman/port.go]: var (
)

File: cmd/podman/port.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	var userPort int
+	var container *libpod.Container
+	var containers []*libpod.Container
+
+	args := c.Args()
+	if err := validateFlags(c, portFlags); err != nil {
+		return err
+	}
+
+	if c.Bool("latest") && c.Bool("all") {
```
Comment (rhatdan) [inline cmd/podman/port.go]: latest is not mentioned above as being supported.

File: cmd/podman/port.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	}
+
+	port := ""
+	if len(args) > 1 && !c.Bool("latest") {
+		port = args[1]
+	}
+	if len(args) == 1 && c.Bool("latest"){
+		port = args[0]
+	}
+	if port != "" {
```
Comment (rhatdan) [inline cmd/podman/port.go]: Can I use port with --all?  If yes then arg[0]would be the port?  Should we just make the port mapping be --port and go away from the Docker default?

File: cmd/podman/port.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+				return errors.Wrapf(err, "unable to format port")
+			}
+
+		}
+		// User supplied port and protocol
+		if len(fields) == 2 {
+			userProto = fields[1]
+		}
+		// Format is incorrect
+		if len(fields) > 2 || len(fields) < 1 {
```
Comment (rhatdan) [inline cmd/podman/port.go]: Could you move this check to right after you split the port?

File: cmd/podman/port.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+		containers, err = runtime.GetRunningContainers()
+		if err != nil {
+			return errors.Wrapf(err, "unable to get all containers")
+		}
+	}
+
+	for _, con := range containers {
+		if state, _ := con.State(); state != libpod.ContainerStateRunning {
+			// If the container is not running, return silently
+			return nil
```
Comment (rhatdan) [inline cmd/podman/port.go]: Shouldn't we just continue.  If I say podman port --all, and a container shuts down between the time I get all and here, we should just ignore the container, not kill off the report,

File: cmd/podman/port.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+			// If not searching by port or port/proto, then dump what we see
+			if port == "" {
+				fmt.Printf("%d/%s -> %s:%d\n", v.ContainerPort, v.Protocol, hostIP, v.HostPort)
+				continue
+			}
+			// We have a match on ports
+			if v.ContainerPort == int32(userPort) {
+				if userProto == "" || userProto == v.Protocol {
+					fmt.Printf("%s:%d", hostIP, v.HostPort)
+					break
```
Comment (rhatdan) [inline cmd/podman/port.go]: Why is this a break?  Shouldn't it continue going through the port mappings?

File: docs/podman-port.1.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+List the port mappings for a specific container for port 80 and the tcp protocol.
+```
+#podman port b4d2f054 80/tcp
+0.0.0.0:44327
+#
+```
+## SEE ALSO
+podman(1)
```
Comment (rhatdan) [inline docs/podman-port.1.md]: Maybe podman-inspect?

File: cmd/podman/port.go (hunk 3 above)
Comment (baude) [reply cmd/podman/port.go]: are you saying you dont see it as a flag ?

File: cmd/podman/port.go (hunk 2 above)
Comment (baude) [reply cmd/podman/port.go]: fixed

File: cmd/podman/port.go (hunk 3 above)
Comment (rhatdan) [reply cmd/podman/port.go]: Oh right, I forgot about your latestflag thing above.  Never mind.

File: cmd/podman/port.go (hunk 5 above)
Comment (baude) [reply cmd/podman/port.go]: done

File: cmd/podman/port.go (hunk 6 above)
Comment (baude) [reply cmd/podman/port.go]: indeed, i originally wrote this without --all so good comment

File: cmd/podman/port.go (hunk 7 above)
Comment (baude) [reply cmd/podman/port.go]: nope, should only be one so it can stop its iteration.

File: docs/podman-port.1.md (hunk 8 above)
Comment (baude) [reply docs/podman-port.1.md]: ok

File: cmd/podman/port.go (hunk 4 above)
Comment (baude) [reply cmd/podman/port.go]: Cannot use port with --all.  

I was trying to keep compatibility but the parsing would be way easier if the port and protocol were called out with --options.

File: cmd/podman/port.go (hunk 1 above)
Comment (baude) [reply cmd/podman/port.go]: fixed

**Comment by baude** [comment]
bot, retest this please

**Comment by baude** [comment]
bot, retest this please

**Comment by baude** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
LGTM

**Comment by baude** [comment]
bot, retest this please

**Comment by rhatdan** [comment]
LGTM
@mheon @umohnani8 @TomSweeneyRedHat PTAL

**Comment by umohnani8** [comment]
LGTM

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Comment by rhatdan** [comment]
Seems like bot is still frozen.  Merging manually.

**Pull request merged**